src/hotspot/share/opto/vectornode.cpp

```diff
@@ -2083,6 +2084,114 @@ Node* VectorBlendNode::Identity(PhaseGVN* phase) {
   return this;
 }
 
+Node* SelectFromTwoVectorNode::Ideal(PhaseGVN* phase, bool can_reshape) {
+  int num_elem = vect_type()->length();
+  BasicType elem_bt = vect_type()->element_basic_type();
+  if (Matcher::match_rule_supported_vector(Op_SelectFromTwoVector, num_elem, elem_bt)) {
+    return nullptr;
+  }
+  Node* index_vec = in(1);
+  Node* src1  = in(2);
+  Node* src2  = in(3);
+
+  // Lower the IR to constituents operations.
+  //   SelectFromTwoVectorNode =
+  //     (VectorBlend
+  //         (VectorRearrange SRC1, INDEX)
+  //         (VectorRearrange SRC2, NORM_INDEX)
+  //         MASK)
+  // This shall prevent an intrinsification failure and associated argument
+  // boxing penalties.
+
+  auto lane_count_type = [&]() {
+    switch(elem_bt) {
+      case T_BYTE:
+      case T_SHORT:
+      case T_INT:
+      case T_FLOAT:
+        return static_cast<const Type*>(TypeInt::make(num_elem));
+      case T_DOUBLE:
+      case T_LONG:
+        return static_cast<const Type*>(TypeLong::make(num_elem));
+      default:
+        fatal("Unsupported vectortype (%s)", type2name(elem_bt));
+        return static_cast<const Type*>(nullptr);
+    }
+  };
+
+  auto make_integral_index_vec = [&](Node* index_vec) {
+    switch(elem_bt) {
+      case T_FLOAT:
+        return phase->transform(new VectorCastF2XNode(index_vec, TypeVect::make(T_INT, num_elem)));
+      break;
+      case T_DOUBLE:
+        return phase->transform(new VectorCastD2XNode(index_vec, TypeVect::make(T_LONG, num_elem)));
+      break;
+      default:
+        return index_vec;
+    }
+  };
+
+  auto get_integal_type = [&](BasicType elem_bt) {
+    switch(elem_bt) {
+      case T_FLOAT:  return T_INT;
+      case T_DOUBLE: return T_LONG;
+      default: return elem_bt;
+    }
+  };
+
+  BasicType integral_elem_bt = get_integal_type(elem_bt);
+  int opc = VectorSupport::vop2ideal(VectorSupport::VECTOR_OP_SUB, integral_elem_bt);
+  int sopc = VectorNode::opcode(opc, integral_elem_bt);
+
+  BoolTest::mask pred = BoolTest::lt;
+  ConINode* pred_node = (ConINode*)phase->makecon(TypeInt::make(pred));
+  Node* lane_cnt = phase->makecon(lane_count_type());
+  Node* bcast_lane_cnt_vec = phase->transform(VectorNode::scalar2vector(lane_cnt, num_elem, Type::get_const_basic_type(integral_elem_bt), false));
+  Node* integral_index_vec = make_integral_index_vec(index_vec);
+
+  // Comparison over integral vectors weeds out emitting additional
+  // instructions for checking special floating point values.
+  const TypeVect* vmask_type = TypeVect::makemask(integral_elem_bt, num_elem);
+  Node* mask = phase->transform(new VectorMaskCmpNode(pred, integral_index_vec, bcast_lane_cnt_vec, pred_node, vmask_type));
+
+  vmask_type = TypeVect::makemask(elem_bt, num_elem);
+  mask = phase->transform(new VectorMaskCastNode(mask, vmask_type));
```

Comment:
I would just have two variables, and not overwrite it: `integral_vmask_type` and `vmask_type`. Maybe also `mask` could be split into two variables?

Reply:
I think the variable names are appropriate and in accordance with convention.